## Упражнения по библиотеке Numpy

In [1]:
import numpy as np

**1.** Дан случайный массив, поменять знак у элементов, значения которых между 3 и 8

In [2]:
arr = np.array([[2,3,4],[6,5,8]])
arr[(arr<8) & (arr>3)] = -1 * (arr[(arr<8) & (arr>3)])
arr

array([[ 2,  3, -4],
       [-6, -5,  8]])

**2.** Заменить максимальный элемент случайного массива на 0

In [3]:
arr = np.array([1,8,3,2,3,4])
arr[np.where(arr == arr.max())] = 0
arr

array([1, 0, 3, 2, 3, 4])

**3.** Построить прямое произведение массивов (все комбинации с каждым элементом). На вход подается двумерный массив

In [4]:
arr1 = np.random.rand(3,3)
arr2 = np.random.rand(3,3)
np.dot(arr1,arr2)

array([[0.77440032, 0.93716085, 1.74801302],
       [0.93111314, 1.06615585, 1.87137024],
       [0.51710799, 0.60145613, 1.13578753]])

**4.** Даны 2 массива A (8x3) и B (2x2). Найти строки в A, которые содержат элементы из каждой строки в B, независимо от порядка элементов в B

In [5]:
###########################################################
A = np.random.randint(5, size=(8,3))
B = np.random.randint(5, size=(2,2))
print(A)
print(B)
A[np.isin(A,B)]

[[1 4 4]
 [3 2 3]
 [0 1 2]
 [4 3 4]
 [4 4 2]
 [0 0 0]
 [0 3 4]
 [2 3 0]]
[[0 2]
 [4 3]]


array([4, 4, 3, 2, 3, 0, 2, 4, 3, 4, 4, 4, 2, 0, 0, 0, 0, 3, 4, 2, 3, 0])

**5.** Дана 10x3 матрица, найти строки из неравных значений (например строка [2,2,3] остается, строка [3,3,3] удаляется)

In [6]:
arr = np.random.randint(5, size=(10,3))
print(arr)
lst = []
for i in range(arr.shape[0]):
    if np.all(arr[i]==arr[i][0]):
        print('Row: ', i)
        lst.append(i)
np.delete(arr, obj=lst, axis=0)

[[2 0 2]
 [1 0 2]
 [4 0 3]
 [4 3 4]
 [0 2 0]
 [2 3 4]
 [0 2 1]
 [2 1 0]
 [1 4 1]
 [3 3 4]]


array([[2, 0, 2],
       [1, 0, 2],
       [4, 0, 3],
       [4, 3, 4],
       [0, 2, 0],
       [2, 3, 4],
       [0, 2, 1],
       [2, 1, 0],
       [1, 4, 1],
       [3, 3, 4]])

**6.** Дан двумерный массив. Удалить те строки, которые повторяются

In [7]:
arr = np.random.randint(3, size=(8,2))
print(arr)
unique = np.unique(arr, axis=0)
unique

[[1 1]
 [0 2]
 [0 0]
 [1 0]
 [1 0]
 [0 1]
 [2 2]
 [1 1]]


array([[0, 0],
       [0, 1],
       [0, 2],
       [1, 0],
       [1, 1],
       [2, 2]])

______
______

Для каждой из следующих задач (1-5) нужно привести 2 реализации – одна без использования numpy (cчитайте, что там, где на входе или выходе должны быть numpy array, будут просто списки), а вторая полностью векторизованная с использованием numpy (без использования питоновских циклов/map/list comprehension).


__Замечание 1.__ Можно считать, что все указанные объекты непустые (к примеру, в __задаче 1__ на диагонали матрицы есть ненулевые элементы).

__Замечание 2.__ Для большинства задач решение занимает не больше 1-2 строк.

___

* __Задача 1__: Подсчитать произведение ненулевых элементов на диагонали прямоугольной матрицы.  
 Например, для X = np.array([[1, 0, 1], [2, 0, 2], [3, 0, 3], [4, 4, 4]]) ответ 3.

In [8]:
# using only python
arr = [[1, 0, 1], [2, 0, 2], [3, 0, 3], [4, 4, 4]]
arr_len = len(arr[0])
count = 1
for i in range(arr_len):
    if arr[i][i] == 0:
        continue
    count *= arr[i][i]
print(count)

3


In [9]:
# using numpy
arr = np.array([[1, 0, 1], [2, 0, 2], [3, 0, 3], [4, 4, 4]])
arr = arr.diagonal()
arr[arr != 0].prod()

3

* __Задача 2__: Даны два вектора x и y. Проверить, задают ли они одно и то же мультимножество.  
  Например, для x = np.array([1, 2, 2, 4]), y = np.array([4, 2, 1, 2]) ответ True.

In [10]:
#using only python
x = [1,2,2,4]
y = [4,2,1,2]
set(x) == set(y)

True

In [11]:
# using numpy
x = np.array([1, 2, 2, 4])
y = np.array([4, 2, 1, 2])
np.isin(x,y).all()

True

* __Задача 3__: Найти максимальный элемент в векторе x среди элементов, перед которыми стоит ноль. 
 Например, для x = np.array([6, 2, 0, 3, 0, 0, 5, 7, 0]) ответ 5.

In [12]:
# using python
x = [6, 2, 0, 3, 0, 0, 5, 7, 0]
max_val = float('-inf')
for i in range(len(x)):
    if i == 0:
        continue
    if x[i] > max_val:
        if x[i-1] == 0:
            max_val = x[i]
print(max_val)

5


In [13]:
# using numpy

* __Задача 4__: Реализовать кодирование длин серий (Run-length encoding). Для некоторого вектора x необходимо вернуть кортеж из двух векторов одинаковой длины. Первый содержит числа, а второй - сколько раз их нужно повторить.  
 Например, для x = np.array([2, 2, 2, 3, 3, 3, 5]) ответ (np.array([2, 3, 5]), np.array([3, 3, 1])).

In [14]:
# using python
x = [2, 2, 2, 3, 3, 3, 5, 6, 6, 8]
last = 0
count = 0
nums = []
nums2 = []
repeat = 0
for i in range(len(x)):
    current = x[i]
    if last != current:
        nums.append(current)
        # append the second list
        if count > 0:
            nums2.append(count)
        count = 0
        last = current
    count += 1
nums2.append(count)

print((nums, nums2))

([2, 3, 5, 6, 8], [3, 3, 1, 2, 1])


In [15]:
# using numpy
# shifting our list and finding where it is != it's value; x[0] - is our first value in list
switches = np.nonzero(np.append(x, 0) != np.append(x[0], x))[0]
repeat = (np.append(switches, switches[-1]) - np.append(0, switches))[0:-1]
print((np.unique(x), repeat))

(array([2, 3, 5, 6, 8]), array([3, 3, 1, 2, 1], dtype=int64))


* __Задача 5__: Даны две выборки объектов - X и Y. Вычислить матрицу евклидовых расстояний между объектами. Сравните с функцией scipy.spatial.distance.cdist по скорости работы.

In [63]:
# using python
%time
X = [i for i in range(50000)]
Y = [i**2 for i in range(50000)]
dist_squared = sum([(X_i - Y_i)**2 for X_i, Y_i in zip(X, Y)])

CPU times: total: 0 ns
Wall time: 0 ns


In [70]:
# using numpy
%time
X2 = np.array(X)
Y2 = np.array(Y)
dist_squared = np.sum(np.square(X2 - Y2))

CPU times: total: 0 ns
Wall time: 0 ns


In [44]:
from scipy.spatial import distance

In [67]:
arr = list(zip(X,Y))

In [68]:
# using scipy
distance.cdist(arr, arr, 'euclidean')

array([[0.00000000e+00, 1.41421356e+00, 4.47213595e+00, ...,
        2.49970001e+09, 2.49980000e+09, 2.49990000e+09],
       [1.41421356e+00, 0.00000000e+00, 3.16227766e+00, ...,
        2.49970001e+09, 2.49980000e+09, 2.49990000e+09],
       [4.47213595e+00, 3.16227766e+00, 0.00000000e+00, ...,
        2.49970001e+09, 2.49980000e+09, 2.49990000e+09],
       ...,
       [2.49970001e+09, 2.49970001e+09, 2.49970001e+09, ...,
        0.00000000e+00, 9.99950000e+04, 1.99992000e+05],
       [2.49980000e+09, 2.49980000e+09, 2.49980000e+09, ...,
        9.99950000e+04, 0.00000000e+00, 9.99970000e+04],
       [2.49990000e+09, 2.49990000e+09, 2.49990000e+09, ...,
        1.99992000e+05, 9.99970000e+04, 0.00000000e+00]])

_______
________

* #### __Задача 6__: CrunchieMunchies __*__

Вы работаете в отделе маркетинга пищевой компании MyCrunch, которая разрабатывает новый вид вкусных, полезных злаков под названием **CrunchieMunchies**.

Вы хотите продемонстрировать потребителям, насколько полезны ваши хлопья по сравнению с другими ведущими брендами, поэтому вы собрали данные о питании нескольких разных конкурентов.

Ваша задача - использовать вычисления Numpy для анализа этих данных и доказать, что ваши **СrunchieMunchies** - самый здоровый выбор для потребителей.


In [18]:
import numpy as np

1. Просмотрите файл cereal.csv. Этот файл содержит количества калорий для различных марок хлопьев. Загрузите данные из файла и сохраните их как calorie_stats.

In [19]:
calorie_stats = np.loadtxt("./data/cereal.csv", delimiter=",")
calorie_stats

array([ 70., 120.,  70.,  50., 110., 110., 110., 130.,  90.,  90., 120.,
       110., 120., 110., 110., 110., 100., 110., 110., 110., 100., 110.,
       100., 100., 110., 110., 100., 120., 120., 110., 100., 110., 100.,
       110., 120., 120., 110., 110., 110., 140., 110., 100., 110., 100.,
       150., 150., 160., 100., 120., 140.,  90., 130., 120., 100.,  50.,
        50., 100., 100., 120., 100.,  90., 110., 110.,  80.,  90.,  90.,
       110., 110.,  90., 110., 140., 100., 110., 110., 100., 100., 110.])

2. В одной порции CrunchieMunchies содержится 60 калорий. Насколько выше среднее количество калорий у ваших конкурентов?

Сохраните ответ в переменной average_calories и распечатайте переменную в терминале

In [20]:
average_calories = calorie_stats.mean()
print(average_calories - 60)

46.883116883116884


3. Корректно ли среднее количество калорий отражает распределение набора данных? Давайте отсортируем данные и посмотрим.

Отсортируйте данные и сохраните результат в переменной calorie_stats_sorted. Распечатайте отсортированную информацию

In [21]:
calorie_stats_sorted = np.sort(calorie_stats)
calorie_stats_sorted

array([ 50.,  50.,  50.,  70.,  70.,  80.,  90.,  90.,  90.,  90.,  90.,
        90.,  90., 100., 100., 100., 100., 100., 100., 100., 100., 100.,
       100., 100., 100., 100., 100., 100., 100., 100., 110., 110., 110.,
       110., 110., 110., 110., 110., 110., 110., 110., 110., 110., 110.,
       110., 110., 110., 110., 110., 110., 110., 110., 110., 110., 110.,
       110., 110., 110., 110., 120., 120., 120., 120., 120., 120., 120.,
       120., 120., 120., 130., 130., 140., 140., 140., 150., 150., 160.])

4. Похоже, что большинство значений выше среднего. Давайте посмотрим, является ли медиана наиболее корректным показателем набора данных.

Вычислите медиану набора данных и сохраните свой ответ в median_calories. Выведите медиану, чтобы вы могли видеть, как она сравнивается со средним значением.

In [22]:
median_calories = np.median(calorie_stats_sorted)
median_calories

110.0

5. В то время как медиана показывает, что по крайней мере половина наших значений составляет более 100 калорий, было бы более впечатляюще показать, что значительная часть конкурентов имеет более высокое количество калорий, чем CrunchieMunchies.

Рассчитайте различные процентили и распечатайте их, пока не найдете наименьший процентиль, превышающий 60 калорий. Сохраните это значение в переменной nth_percentile.

In [23]:
forth_percentile = np.percentile(calorie_stats, 4)

6. Хотя процентиль показывает нам, что у большинства конкурентов количество калорий намного выше, это неудобная концепция для использования в маркетинговых материалах.

Вместо этого давайте подсчитаем процент хлопьев, в которых содержится более 60 калорий на порцию. Сохраните свой ответ в переменной more_calories и распечатайте его

In [24]:
more_calories = calorie_stats[calorie_stats > 60].shape[0] / calorie_stats.shape[0] * 100
more_calories

96.1038961038961

7. Это действительно высокий процент. Это будет очень полезно, когда мы будем продвигать CrunchieMunchies. Но один вопрос заключается в том, насколько велики различия в наборе данных? Можем ли мы сделать обобщение, что в большинстве злаков содержится около 100 калорий или разброс еще больше?

Рассчитайте величину отклонения, найдя стандартное отклонение, Сохраните свой ответ в calorie_std и распечатайте на терминале. Как мы можем включить эту ценность в наш анализ?

In [25]:
calorie_std = calorie_stats.std()
calorie_std

19.35718533390827

8. Напишите короткий абзац, в котором кратко изложите свои выводы и то, как, по вашему мнению, эти данные могут быть использованы в интересах Mycrunch при маркетинге CrunchieMunchies.

Поскольку мы выяснили, что у 96% конкурентов число содержания калорий в злаках выше, это поможет позиционировать продукт как содержащий мало сахара и его можно будет отнести к нише здорового питания. Однако чтобы делать такие выводы, нам необходимо убедиться в корректности нашей выборки, необходимо чтобы данная выборка являлась рандомизированной и чтобы она соответствовала генеральной совокупности.